### DISCUSSION TITLE
scss compiler and yoga

### DISCUSSION THREAD

**Discussion by cathaysia** [mikke89/RmlUi#375] [Ideas]
As a superset of CSS, scss is widely used in front-end development. Inheriting the scss compiler for the program can greatly bring the pleasure of the development phase. Although libscss development has stagnated, I think its existing functionality is sufficient.
Another problem is yoga. Yoga is a flexbox layout engine written by facebook. The performance and stability are also reliable. I think it seems better to use existing mature engine libraries.

I inherited the above two libraries into qt some time ago, and they performed well. But I don't know if rmlui will accept these changes. Both libraries above can be included directly as source code. The workload is not too big.

I want to hear your thoughts. Are these two jobs required? The first job is much simpler and can be completed quickly, but the second job needs to be fully tested.

- https://github.com/sass/libsass
- https://github.com/facebook/yoga

**Comment by mikke89** [comment]
Hey, and thanks for the suggestions.

Generally, I want to say that we try to reduce the number of dependencies and reduce bloat as much as possible. Both of these libraries duplicate work that is already being done in the library. The yoga library in particular makes little sense as we already have flexbox support.

I took a closer look at LibSass. In my opinion, this is too big to include directly. If users want this feature they can integrate it on the client-side and hand over the compiled CSS to RmlUi. It doesn't really need any changes on the library side.

With that said, I'm open to contributions that extend our own CSS parser and include some SCSS features that we find particularly useful.

**Comment by cathaysia** [comment]
Yes. I can totally understand your consideration. Fewer dependencies and lighter code is undoubtedly very good. But have you considered integrating external libraries directly in the form of copying source code?
By copying the source code to rely on, no external dependencies will be generated, and the size of dependencies will be significantly reduced. For example libsass has a source dependency of 1.7MB. And the source code dependency of yoga is 332 KB.

Based on this inference, we can actually include a lightweight library such as lua as a built-in function. This reduces external dependencies to a certain extent, and on the other hand provides more out-of-the-box functions

**Reply by mikke89** [reply]
I'm more worried about bloat in terms of increased binary size, compilation times, duplicate functionality, runtime performance, and added complexity. Exactly how the thirdparty library is integrated doesn't really matter in terms of these metrics.

Besides the added bloat, there is also the consideration of writing the integration code and maintaining it. The incentives to add these two libraries in particular are not compelling enough to overcome these costs in my opinion, especially in light of the alternatives I suggested in the previous post.

**Reply by cathaysia** [reply]
emm, understand, thanks.
